sharktuner/sharktuner/spec_builder.py

```diff
@@ -17,170 +20,330 @@
-            op.opview.attributes[ROOT_OP_ATTR_NAME], ir.UnitAttr
-        ), f"expected '{ROOT_OP_ATTR_NAME}' attr to be a unit attr"
-    if has_root_attr:
-        del op.opview.attributes[ROOT_OP_ATTR_NAME]
-    # Get the root op string for formatting the final spec.
-    root_operation = str(op)
-    if has_root_attr:
-        op.opview.attributes[ROOT_OP_ATTR_NAME] = ir.UnitAttr.get(op.context)
-
-    if linalg.isa_contraction_op(op):
-        # Temporary solution using custom contraction transform ops for contraction operations.
-        inputs = op.opview.operands
-        outputs = op.opview.results
-        lhs_type = str(ir.RankedTensorType(inputs[0].type).element_type)
-        rhs_type = str(ir.RankedTensorType(inputs[1].type).element_type)
-        output_type = str(ir.RankedTensorType(outputs[0].type).element_type)
-
-        contraction_dims = linalg.infer_contraction_dimensions(op)
-        indexing_maps = linalg.get_indexing_maps(op)
-        maps = [map_attr.value for map_attr in indexing_maps]
-        lhs_dims = common.get_map_result_dim_positions(maps[0])
-        rhs_dims = common.get_map_result_dim_positions(maps[1])
-        assert lhs_dims, "no lhs dimensions"
-        assert rhs_dims, "no rhs dimensions"
-        lhs_type_shape = ir.RankedTensorType(op.operands[0].type)
-        rhs_type_shape = ir.RankedTensorType(op.operands[1].type)
-
-        m_dims = [
-            lhs_type_shape.shape[lhs_dims.index(dim)] for dim in contraction_dims.m
-        ]
-        n_dims = [
-            rhs_type_shape.shape[rhs_dims.index(dim)] for dim in contraction_dims.n
-        ]
-        k_dims = [
-            lhs_type_shape.shape[lhs_dims.index(dim)] for dim in contraction_dims.k
-        ]
-        batch_dims = [
-            lhs_type_shape.shape[lhs_dims.index(dim)] for dim in contraction_dims.batch
-        ]
-
-        dims_equal_checks = []
-        dims_equal_checks.append(
-            f"transform.iree.match.dims_equal %batch, {batch_dims} : !transform.param<i64>"
-        )
-        dims_equal_checks.append(
-            f"transform.iree.match.dims_equal %m, {m_dims} : !transform.param<i64>"
-        )
-        dims_equal_checks.append(
-            f"transform.iree.match.dims_equal %n, {n_dims} : !transform.param<i64>"
+class SpecBuilder(ABC):
+    def __init__(self, opinfo: OpInfo):
+        self.opinfo = opinfo
+
+    def create_config_params(
+        self, config_list: list[common.TuningConfiguration]
+    ) -> list[ir.Value]:
+        """
+        Creates parameter constant operations for each configuration.
+        """
+        config_params = []
+        for config in config_list:
+            config_param = transform.ParamConstantOp(
+                transform.AnyParamType.get(),
+                config.configuration,
+            ).result
+            config_params.append(config_param)
+        return config_params
+
+    @staticmethod
+    def get_placeholder_spec(context: ir.Context) -> ir.Module:
```

Comment:
What's the benefit of building the base spec programmatically? If it's going to be the same for all inputs, having hardocded mlir sounds OK to me and is much more concise.